src/main/kotlin/com/demonwav/mcdev/i18n/i18n-utils.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Minecraft Dev for IntelliJ
+ *
+ * https://minecraftdev.org
+ *
+ * Copyright (c) 2017 minecraft-dev
+ *
+ * MIT License
+ */
+
+package com.demonwav.mcdev.i18n
+
+import com.demonwav.mcdev.i18n.lang.I18nFile
+import com.demonwav.mcdev.i18n.lang.I18nFileType
+import com.demonwav.mcdev.i18n.lang.gen.psi.I18nProperty
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.PsiManager
+import com.intellij.psi.search.FileTypeIndex
+import com.intellij.psi.search.GlobalSearchScope
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.util.indexing.FileBasedIndex
+
+enum class Scope {
+    GLOBAL, PROJECT
+}
+
+private fun Project.files(scope: Scope) =
+    FileBasedIndex.getInstance().getContainingFiles(FileTypeIndex.NAME, I18nFileType, if (scope == Scope.GLOBAL) GlobalSearchScope.allScope(this) else GlobalSearchScope.projectScope(this))
+        .mapNotNull { PsiManager.getInstance(this).findFile(it) as I18nFile? }
+
+private fun Project.findPropertiesImpl(scope: Scope, fileFilter: (I18nFile) -> Boolean = { true }, propertyFilter: (I18nProperty) -> Boolean = { true }) =
+    files(scope)
+        .filter(fileFilter)
```

Comment:
Considering you do 3 separate operations that require copying all the contents to a new list you may want to consider using Kotlin sequences to do it in one run (or a Java stream)